internal/controllers/machineset/machineset_preflight.go

```diff
@@ -0,0 +1,217 @@
+	if err != nil {
+		if errors.Is(err, contract.ErrFieldNotFound) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to get the version of ControlPlane %s", action, cpKlogRef)
+	}
+	cpSemver, err := semver.ParseTolerant(*cpVersion)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to parse version %q of ControlPlane %s", action, *cpVersion, cpKlogRef)
+	}
+
+	result := ctrl.Result{}
+	errList := []error{}
+
+	// Run the control-plane-stable preflight check.
+	if !skipped.Has(clusterv1.MachineSetPreflightCheckControlPlaneStable) {
+		if checkResult, err := r.controlPlaneStablePreflightCheck(ctx, controlPlane, action); err != nil || !checkResult.IsZero() {
+			result = util.LowestNonZeroResult(result, checkResult)
+			if err != nil {
+				errList = append(errList, err)
+			}
+		}
+	}
+
+	// Check the version skew policies only if version is defined in the MachineSet.
+	if ms.Spec.Template.Spec.Version != nil {
+		msVersion := *ms.Spec.Template.Spec.Version
+		msSemver, err := semver.ParseTolerant(msVersion)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform preflight checks: failed to parse version %q of MachineSet %s", action, msVersion, klog.KObj(ms))
+		}
+
+		// Run the kubernetes-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubernetesVersion) {
+			if checkResult := r.kubernetesVersionPreflightCheck(ctx, cpSemver, msSemver, action); !checkResult.IsZero() {
+				result = util.LowestNonZeroResult(result, checkResult)
+			}
+		}
+
+		// Run the kubeadm-version skew preflight check.
+		if !skipped.Has(clusterv1.MachineSetPreflightCheckKubeadmVersion) {
+			if checkResult, err := r.kubeadmVersionPreflightCheck(ctx, cpSemver, msSemver, ms, action); err != nil || !checkResult.IsZero() {
+				result = util.LowestNonZeroResult(result, checkResult)
+				if err != nil {
+					errList = append(errList, err)
+				}
+			}
+		}
+	}
+
+	if len(errList) > 0 {
+		return ctrl.Result{}, errors.Wrapf(kerrors.NewAggregate(errList), "failed to perform %q: failed to perform preflight checks", action)
+	}
+	return result, nil
+}
+
+func (r *Reconciler) controlPlaneStablePreflightCheck(ctx context.Context, controlPlane *unstructured.Unstructured, action string) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// Check that the control plane is not provisioning.
+	isProvisioning, err := contract.ControlPlane().IsProvisioning(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q: failed to perform %q preflight check: failed to check if ControlPlane %s is provisioning", action, clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef)
```

Comment:
```suggestion
		return ctrl.Result{}, errors.Wrapf(err, "failed to perform %q preflight check: failed to check if ControlPlane %s is provisioning", action, clusterv1.MachineSetPreflightCheckControlPlaneStable, cpKlogRef)
```
It's redundant in the errors as we already add the action one layer above when we wrap

(same in other places below, we should keep it in the logs though)